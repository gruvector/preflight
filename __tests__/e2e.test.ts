import execa from 'execa';
import pMap from 'p-map';

const fixturesTempDir = '__tests__/fixtures/__temp';

async function cloneRepoToFixtures(repoPath: string, fixtureDirName: string) {
  return execa.command(
    `git clone --depth 1 --single-branch --branch=main https://github.com/${repoPath}.git ${fixturesTempDir}/${fixtureDirName} --config core.autocrlf=input`,
  );
}

type Repo = {
  repoPath: string;
  dirName: string;
  installCommands?: string[];
};

const testRepos: Repo[] = [
  {
    repoPath: 'upleveled/preflight-test-project-react-passing',
    dirName: 'react-passing',
    // To use install commands, for example to install or upgrade
    // a package first, pass an installCommands array like this:
    // installCommands: [
    //   // To install the latest version of the ESLint config
    //   'yarn upgrade --latest @upleveled/eslint-config-upleveled',
    //   // Avoid any issues with uncommitted files
    //   'git reset --hard HEAD',
    // ],
  },
];

beforeAll(
  async () => {
    await pMap(
      testRepos,
      async ({ repoPath, dirName }) => cloneRepoToFixtures(repoPath, dirName),
      { concurrency: 4 },
    );

    await pMap(
      testRepos,
      async ({ dirName, installCommands }) => {
        if (!installCommands || installCommands.length < 1) {
          return execa.command('yarn --frozen-lockfile', {
            cwd: `${fixturesTempDir}/${dirName}`,
          });
        }

        return pMap(
          installCommands,
          async (command) =>
            execa.command(command, {
              cwd: `${fixturesTempDir}/${dirName}`,
            }),
          { concurrency: 1 },
        );
      },
      { concurrency: 1 },
    );
  },
  // 5 minute timeout for Yarn installation inside test repos
  300000,
);

test('Passes in the react-passing test project', async () => {
  const { stdout, stderr } = await execa.command(
    `../../../../bin/preflight.js`,
    {
      cwd: `${fixturesTempDir}/react-passing`,
    },
  );

  const stdoutSortedWithoutVersionNumber = stdout
    .replace(/(UpLeveled Preflight) v\d+\.\d+\.\d+/, '$1')
    .split('\n')
    .sort((a: string, b: string) => {
      if (b.includes('UpLeveled Preflight')) return 1;
      return a < b ? -1 : 1;
    })
    .join('\n');

  expect(stdoutSortedWithoutVersionNumber).toMatchSnapshot();
  expect(stderr.replace(/^\(node:\d+\) /, '')).toMatchSnapshot();
}, 30000);
