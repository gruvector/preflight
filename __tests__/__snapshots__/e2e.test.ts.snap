// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Passes in the react-passing test project 1`] = `
"ðŸš€ UpLeveled Preflight
[STARTED] All changes committed to Git
[STARTED] ESLint
[STARTED] ESLint config is latest version
[STARTED] GitHub repo has deployed project link under About
[STARTED] No dependencies without types
[STARTED] No dependency problems
[STARTED] No extraneous files committed to Git
[STARTED] No secrets committed to Git
[STARTED] No unused dependencies
[STARTED] Preflight is latest version
[STARTED] Prettier
[STARTED] Project folder name matches correct format
[STARTED] Use single package manager
[STARTED] node_modules/ folder ignored in Git
[SUCCESS] All changes committed to Git
[SUCCESS] ESLint
[SUCCESS] ESLint config is latest version
[SUCCESS] GitHub repo has deployed project link under About
[SUCCESS] No dependencies without types
[SUCCESS] No dependency problems
[SUCCESS] No extraneous files committed to Git
[SUCCESS] No secrets committed to Git
[SUCCESS] No unused dependencies
[SUCCESS] Preflight is latest version
[SUCCESS] Prettier
[SUCCESS] Project folder name matches correct format
[SUCCESS] Use single package manager
[SUCCESS] node_modules/ folder ignored in Git"
`;

exports[`Passes in the react-passing test project 2`] = `""`;
